lib/ex_unit/lib/ex_unit/on_exit_handler.ex

```diff
@@ -1,39 +1,75 @@
 defmodule ExUnit.OnExitHandler do
   @moduledoc false
+
   @name __MODULE__
+  @supervisor 2
+  @on_exit 3
 
   use Agent
 
+  @spec start_link(keyword()) :: {:ok, pid}
   def start_link(_opts) do
-    Agent.start_link(fn -> %{} end, name: @name)
+    Agent.start_link(fn ->
+      :ets.new(@name, [:public, :named_table, read_concurrency: true, write_concurrency: true])
+    end, name: @name)
   end
 
   @spec register(pid) :: :ok
   def register(pid) when is_pid(pid) do
-    Agent.update(@name, &Map.put(&1, pid, []))
+    :ets.insert(@name, {pid, nil, []})
   end
 
   @spec add(pid, term, (() -> term)) :: :ok | :error
   def add(pid, name_or_ref, callback) when is_pid(pid) and is_function(callback, 0) do
-    Agent.get_and_update(@name, fn map ->
-      if entries = Map.get(map, pid) do
+    try do
+      :ets.lookup_element(@name, pid, @on_exit)
+    rescue
+      _ -> :error
+    else
+      entries ->
         entries = List.keystore(entries, name_or_ref, 0, {name_or_ref, callback})
-        {:ok, Map.put(map, pid, entries)}
-      else
-        {:error, map}
-      end
-    end)
+        true = :ets.update_element(@name, pid, {@on_exit, entries})
+        :ok
+    end
+  end
+
+  @spec get_supervisor(pid) :: {:ok, pid | nil} | :error
+  def get_supervisor(pid) when is_pid(pid) do
+    try do
+      {:ok, :ets.lookup_element(@name, pid, @supervisor)}
+    rescue
+      _ -> :error
+    end
+  end
+
+  @spec put_supervisor(pid, pid) :: :ok | :error
+  def put_supervisor(pid, sup) when is_pid(pid) and is_pid(sup) do
+    case :ets.update_element(@name, pid, {@supervisor, sup}) do
+      true -> :ok
+      false -> :error
+    end
   end
 
   @spec run(pid, timeout) :: :ok | {Exception.kind, term, Exception.stacktrace}
   def run(pid, timeout) when is_pid(pid) do
-    callbacks = Agent.get_and_update(@name, &Map.pop(&1, pid, []))
-    exec_on_exit_callbacks(Enum.reverse(callbacks), timeout)
+    [{^pid, sup, callbacks}] = :ets.take(@name, pid)
+    error = terminate_supervisor(sup, timeout)
+    exec_on_exit_callbacks(Enum.reverse(callbacks), timeout, error)
+  end
+
+  defp terminate_supervisor(nil, _timeout), do: nil
+  defp terminate_supervisor(sup, timeout) do
+    ref = Process.monitor(sup)
+    receive do
+      {:DOWN, ^ref, _, _, _} -> nil
+    after
+      timeout -> ExUnit.TimeoutError.exception(timeout: timeout, type: :supervised)
```

Comment:
This will return `Exception.t` but we want `{:error, Exception.t, stacktrace()}` - this is no test for this.